File: srgb2lin.py
# Made because PVRTexToolCLI sucks
# EmK530 2023-07-04

import math
from PIL import Image

def convert(path):
    def srgb2lin(s):
        if s <= 0.0404482362771082:
            lin = s / 12.92
        else:
            lin = pow(((s + 0.055) / 1.055), 2.4)
        return lin
    im=Image.open(path)
    new=[]
    for i in im.getdata():
        if len(i) == 3:
            new.append((
                math.floor(srgb2lin(i[0])/2058.61501702),
                math.floor(srgb2lin(i[1])/2058.61501702),
                math.floor(srgb2lin(i[2])/2058.61501702),
            ))
        else:
            new.append((
                math.floor(srgb2lin(i[0])/2058.61501702),
                math.floor(srgb2lin(i[1])/2058.61501702),
                math.floor(srgb2lin(i[2])/2058.61501702),
                i[3]
            ))
    im.close()
    newim=Image.new(im.mode,im.size)
    newim.putdata(new)
    newim.save(path)

File: srgb2lin.cs
// Script made by zuzaratrust and EmK530

using SixLabors.ImageSharp;
using SixLabors.ImageSharp.PixelFormats;

public static class srgb2lin
{
    static byte[] preComputed = new byte[256];
    static double factor = 1.0 / 2058.61501702;

    static double tolin(int s)
    {
        double lin;
        if (s < 11)
        {
            lin = (double)s / 12.92;
        }
        else
        {
            lin = Math.Pow(((double)s + 0.055) / 1.055, 2.4);
        }
        return lin;
    }

    static bool computed = false;

    static void preCompute()
    {
        for(int i = 0; i < 256; i++)
        {
            preComputed[i] = (byte)Math.Floor(tolin(i) * factor);
        }
        computed = true;
    }

    public static void convert(string outPath)
    {
        if (!computed)
        {
            preCompute();
        }

        Image<Rgba32> img = Image.Load<Rgba32>(outPath);

        for (int y = 0; y < img.Height; y++)
        {
            for (int x = 0; x < img.Width; x++)
            {
                Rgba32 px = img[x, y];
                px.R = preComputed[px.R];
                px.G = preComputed[px.G];
                px.B = preComputed[px.B];
                img[x, y] = px;
            }
        }

        img.Save(outPath);
    }
}
